packages/contracts-bedrock/scripts/DeployImplementations.s.sol

```diff
@@ -0,0 +1,325 @@
+    //   - MIPS: Not proxied, shared by all standard chains.
+    //   - OptimismPortal2: Proxied, shared by all standard chains.
+    //
+    // This script only deploys the shared contracts. The bespoke contracts are deployed by
+    // `DeployOPChain.s.sol`. When the shared contracts are proxied, we call the "implementations",
+    // and when they are not proxied, we call them "singletons". So here we deploy:
+    //
+    //   - OptimismPortal2 (implementation)
+    //   - DelayedWETH (implementation)
+    //   - PreimageOracle (singleton)
+    //   - MIPS (singleton)
+
+    function deployOptimismPortalImpl(DeployImplementationsInput _dsi, DeployImplementationsOutput _dso) public {
+        uint256 proofMaturityDelaySeconds = _dsi.proofMaturityDelaySeconds();
+        uint256 disputeGameFinalityDelaySeconds = _dsi.disputeGameFinalityDelaySeconds();
+
+        vm.broadcast(msg.sender);
+        OptimismPortal2 optimismPortal2Impl = new OptimismPortal2({
+            _proofMaturityDelaySeconds: proofMaturityDelaySeconds,
+            _disputeGameFinalityDelaySeconds: disputeGameFinalityDelaySeconds
+        });
+
+        vm.label(address(optimismPortal2Impl), "OptimismPortal2Impl");
+        _dso.set(_dso.optimismPortal2Impl.selector, address(optimismPortal2Impl));
+    }
+
+    function deployDelayedWETHImpl(DeployImplementationsInput _dsi, DeployImplementationsOutput _dso) public {
+        uint256 withdrawalDelaySeconds = _dsi.withdrawalDelaySeconds();
+
+        vm.broadcast(msg.sender);
+        DelayedWETH delayedWETHImpl = new DelayedWETH({ _delay: withdrawalDelaySeconds });
+
+        vm.label(address(delayedWETHImpl), "DelayedWETHImpl");
+        _dso.set(_dso.delayedWETHImpl.selector, address(delayedWETHImpl));
+    }
+
+    function deployPreimageOracleSingleton(DeployImplementationsInput _dsi, DeployImplementationsOutput _dso) public {
+        uint256 minProposalSizeBytes = _dsi.minProposalSizeBytes();
+        uint256 challengePeriodSeconds = _dsi.challengePeriodSeconds();
+
+        vm.broadcast(msg.sender);
+        PreimageOracle preimageOracleSingleton =
+            new PreimageOracle({ _minProposalSize: minProposalSizeBytes, _challengePeriod: challengePeriodSeconds });
+
+        vm.label(address(preimageOracleSingleton), "PreimageOracleSingleton");
+        _dso.set(_dso.preimageOracleSingleton.selector, address(preimageOracleSingleton));
+    }
+
+    function deployMipsSingleton(DeployImplementationsInput, DeployImplementationsOutput _dso) public {
+        IPreimageOracle preimageOracle = IPreimageOracle(_dso.preimageOracleSingleton());
+
+        vm.broadcast(msg.sender);
+        MIPS mipsSingleton = new MIPS(preimageOracle);
+
+        vm.label(address(mipsSingleton), "MIPSSingleton");
+        _dso.set(_dso.mipsSingleton.selector, address(mipsSingleton));
+    }
+
+    // -------- Utilities --------
+
+    function toIOAddress(address _sender, string memory _identifier) internal pure returns (address) {
```

Comment:
This can be part of a library

Reply:
Also done in https://github.com/ethereum-optimism/optimism/pull/11555